app/scheduled/app-client-expire/index.ts

```diff
@@ -0,0 +1,113 @@
+import dedent from 'ts-dedent'
+
+import {
+  deleteAppClient,
+  extractAttributeRequired,
+  getCognitoUserFromSub,
+} from '~/lib/cognito.server'
+import { sendEmail, sendEmailBulk } from '~/lib/email.server'
+import { feature, origin } from '~/lib/env.server'
+import type { RedactedClientCredential } from '~/routes/user.credentials/client_credentials.server'
+
+export async function handler() {
+  if (!feature('APP_CLIENT_TRACKING')) return
+  const db = await tables()
+
+  const deletionDate = new Date()
+  const deletionCutoff = deletionDate.setMonth(deletionDate.getMonth() - 1)
+
+  const warningDate = new Date()
+  const warningCutoff = warningDate.setDate(warningDate.getDate() - 21)
+
+  // Check for all credentials that should be expired today
+  const expiredCredentials = (
+    await db.client_credentials.query({
+      FilterExpression: 'lastUsed < :deletionCutoff',
+      ExpressionAttributeValues: {
+        ':deletionCutoff': deletionCutoff,
+      },
+    })
+  ).Items as ({ sub: string } & RedactedClientCredential)[]
+
+  await Promise.all([
+    // Mark as expired in DynamoDB
+    ...expiredCredentials.map((cred) =>
+      db.client_credentials.update({
+        Key: { sub: cred.sub, client_id: cred.client_id },
+        UpdateExpression: 'SET expired = :expired',
+        ExpressionAttributeValues: {
+          ':expired': Date.now(),
+        },
+      })
+    ),
+    // Delete App clients from Cognito
+    ...expiredCredentials.map((cred) => deleteAppClient(cred.client_id)),
+  ])
+
+  const usersMapForExpiredClients = Object.fromEntries(
+    await Promise.all(
+      expiredCredentials.map(async (credential) => [
+        credential.name,
+        extractAttributeRequired(
+          (await getCognitoUserFromSub(credential.sub)).Attributes,
+          'email'
+        ),
+      ])
+    )
+  )
+  const expirationEmailPromises = expiredCredentials.map((cred) =>
+    sendEmail({
+      to: [usersMapForExpiredClients[cred.name]],
+      fromName: 'GCN Kafka Service',
+      subject: 'Client Credential Deleted',
+      body: `Your Kafka client credential "${cred.name}" has expired. For more information about our credential expiration policy, please visit ${origin}/docs/internal/auth or ${origin}/contact for support.`,
+    })
+  )
+
+  // Check for all credentials that are going to be expired in the near future
+  const warningCredentials = (
+    await db.client_credentials.query({
+      FilterExpression:
+        'lastUsed < :warningCutoff and attribute_not_exists(expired)',
+      ExpressionAttributeValues: {
+        ':warningCutoff': warningCutoff,
+      },
+    })
+  ).Items as ({ sub: string } & RedactedClientCredential)[]
+
+  const emailsForWarnings = await Promise.all(
+    warningCredentials.map(async (credential) =>
+      extractAttributeRequired(
+        (await getCognitoUserFromSub(credential.sub)).Attributes,
+        'email'
+      )
+    )
+  )
+
+  const warningEmailPromise = sendEmailBulk({
```

Comment:
Rather than sending a bulk email, can we send individual emails for each user so that they specifically describe the credential that is expiring?